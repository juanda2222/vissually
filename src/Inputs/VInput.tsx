import React, {
  useContext,
  useState,
  //useEffect,
  //useRef,
} from 'react'


import styles from './VInput.module.css'
import styled from 'styled-components';
import { Color2Vec } from "../Tools/ColorTools"
import { DefaultThemes, ThemeContext } from "../ThemeProvider/ThemeProvider"


const StyledInput = styled("input") <{ rgbTextColor: number[], rgbFocusedColor: number[] }>`
  border: 2px solid ${props => `rgb(${props.rgbTextColor[0] -30}, ${props.rgbTextColor[1] -30}, ${props.rgbTextColor[2] -30})`};
  color: ${props => `rgb(${props.rgbTextColor[0]}, ${props.rgbTextColor[1]}, ${props.rgbTextColor[2]})`};
  &:focus {
    border-color: ${props => `rgb(${props.rgbFocusedColor[0]}, ${props.rgbFocusedColor[1]}, ${props.rgbFocusedColor[2]});`};
  }
   
`;

const StyledLabel = styled("label") <{
  isFocused: boolean,
  rgbTextColor: number[],
  rgbFocusedTextColor: number[]
}>`
  color: ${props => props.isFocused ?
    `rgb(${props.rgbFocusedTextColor[0]}, ${props.rgbFocusedTextColor[1]}, ${props.rgbFocusedTextColor[2]})` :
    `rgb(${props.rgbTextColor[0]}, ${props.rgbTextColor[1]}, ${props.rgbTextColor[2]})`
  };
`;

function VTextInput ({
  //styling properties
  className,
  style,
  parentStyle,
  labelStyle,
  theme,
  primary, secondary, primaryDark, secondaryDark,

  //functional properties
  value,
  id,
  name,
  type,
  required,
  placeholder,
  onChange,
  onFocus,
  onBlur,
}: InputProps ){  

  // get the theme from the context
  const context_theme = useContext(ThemeContext);
  const [inputIsFocused, setInputFocused] = useState(false)

  // all the event handlers wrapped for custom styling
  const wrappedOnFocus = (
    event: React.FocusEvent<HTMLInputElement>,
  ) => {
    setInputFocused(true)
    if (typeof onFocus === 'function') {onFocus(event)}
  }

  const wrappedOnBlur = (
    event: React.FocusEvent<HTMLInputElement>,
  ) => {
    setInputFocused(false)
    if (typeof onBlur === 'function') {onBlur(event)}
  }

  // generate the theme depending on the boolean inputs or the theme input
  const wrappedClassName = [styles.input_element, className ? className : ""].join(" ")
  let current_theme: DefaultTheme;
  if (secondary) {
    current_theme = DefaultThemes.secondary
  } else if (primaryDark) {
    current_theme = DefaultThemes.primaryDark
  } else if (secondaryDark) {
    current_theme = DefaultThemes.secondaryDark
  } else if (primary) {
    current_theme = DefaultThemes.primary
  } else if (!(typeof (context_theme) == "undefined")) {
    current_theme = context_theme
  } else {
    current_theme = {
      ...DefaultThemes.primary,
      ...(theme && theme)
    } 
  }
  
  // generate the component from the style
  const rgbMainColor = Color2Vec(current_theme.color1)
  const rgbTextColor = Color2Vec(current_theme.textColor1)
  
  return (
    <div
      className={styles.container}
      style={parentStyle}
    >
      <StyledInput
        //custom style:
        rgbFocusedColor={rgbMainColor}
        rgbTextColor={rgbTextColor}
        // parent styling props:
        className={wrappedClassName}
        style={style}
        // functional props:
        value={value}
        placeholder={placeholder}
        id={id}
        name={name}
        type={type}
        required={required}
        onChange={onChange}
        onFocus={wrappedOnFocus}
        onBlur={wrappedOnBlur}
      />      
      <StyledLabel
        className={styles.label_element}
        style={labelStyle}
        isFocused={inputIsFocused}
        rgbTextColor={rgbTextColor}
        rgbFocusedTextColor={rgbMainColor}
        
      >
        {placeholder}
      </StyledLabel>
    </div>
  )
}


export default VTextInput