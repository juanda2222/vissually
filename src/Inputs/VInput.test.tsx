
import React from "react"
import renderer from 'react-test-renderer';
import VInput from './VInput'

describe('VInput', () => {

  test('VInput declaration has no errors', () => {
    expect(VInput).toBeTruthy()
  })

  test("VInput dDefault render. Snapshot test", () => { 
    const TextInputTree = renderer.create(<VInput/>).toJSON();
    console.log(TextInputTree)
    expect(TextInputTree).toMatchSnapshot();
  })

  test("VInput rendering with props. Snapshot test", () => { 

    const TextInputTreeProps = renderer.create(
        <VInput
        className="some_classname"
        placeholder="heeey"
        onChange={(e) => {
          console.log("on change")
          console.log("Value", e.target.value)
        }}
        />
    ).toJSON();

    console.log(TextInputTreeProps)
    expect(TextInputTreeProps).toMatchSnapshot();
  })

  test("VInput rendering with props type='number'. Snapshot test", () => { 

    const TextInputTreeProps = renderer.create(
        <VInput
        className="some_classname"
        placeholder="age"
        name="age"
        type="number"
        onChange={(e) => {
          console.log("on change")
          console.log("Value", e.target.value)
        }}
        />
    ).toJSON();

    console.log(TextInputTreeProps)
    expect(TextInputTreeProps).toMatchSnapshot();
  })


})
