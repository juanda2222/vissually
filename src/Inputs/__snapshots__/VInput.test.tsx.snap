// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VInput VInput dDefault render. Snapshot test 1`] = `
<div
  className="container"
>
  <input
    className="sc-AxjAm LIrVh input_element "
    onBlur={[Function]}
    onFocus={[Function]}
  />
  <label
    className="sc-AxirZ iyFEor label_element"
  />
</div>
`;

exports[`VInput VInput rendering with props type='number'. Snapshot test 1`] = `
<div
  className="container"
>
  <input
    className="sc-AxjAm LIrVh input_element some_classname"
    name="age"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="age"
    type="number"
  />
  <label
    className="sc-AxirZ iyFEor label_element"
  >
    age
  </label>
</div>
`;

exports[`VInput VInput rendering with props. Snapshot test 1`] = `
<div
  className="container"
>
  <input
    className="sc-AxjAm LIrVh input_element some_classname"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="heeey"
  />
  <label
    className="sc-AxirZ iyFEor label_element"
  >
    heeey
  </label>
</div>
`;
